Add support for Laravel-style relations

bennothommo opened the pull request:
This PR allows people to define Laravel-style relations using methods that return `Relation` objects. So instead of:

```php
public $belongsTo = [
     'user' => User::class,
];
```

You may use:

```php
public function user(): BelongsTo
{
    return $this->belongsTo(User::class);
}
```

The main reason for this is to reduce friction for porting Laravel code over to Winter, and as a preference point for people who may wish to use the more declarative style of defining relations.

I've adjusted the unit tests to demonstrate this in action, and will be porting over the relation tests in the Winter `system` module into Storm, as they really should be in Storm anyway.

In order to maintain parity with the array-style configuration of relations in Winter, method-defined relations will need to either implement one of the following to indicate that they provide relations, and allow certain functionality to find them as required (for example, deleting related models when deleting a primary model):

- The relation method must have a return type that correlates to a `Relation` class (like with the example above)
- Alternatively, the method must use the `Winter\Storm\Database\Attributes\Relation` attribute and define the relation type, like so:

```php
use Winter\Storm\Database\Attributes\Relation;

#[Relation('belongsTo')]
public function user()
{
    return $this->belongsTo(User::class);
}
```

### Todo

- [x] Fix attachment relations and tests
- [x] Add support for model traits with new format
- [x] Finish porting tests from Database Tester plugin in core
- [x] Full documentation of new syntax

bennothommo committed:
WIP Laravel style relations

bennothommo committed:
Discard filesystem test case.

We shouldn't test completely third-party code.

bennothommo committed:
Add BelongsToMany, HasMany and HasOne tests

bennothommo committed:
Remove unneeded relation class.

We no longer extend this class, and it doesn't seem to provide any functionality anyway.

bennothommo committed:
Add HasManyThrough, HasOneThrough, MorphMany and MorphOne tests

bennothommo committed:
Move performDeleteOnRelations method into HasRelationship concern

This method is strictly a relationship method and should be grouped with the other relationship methods.

bennothommo committed:
Significantly refactor relation construction.

In order to maintain some semblance of parity with Laravel, we'll now use Laravel's relation constructors and simply overwrite the "new" relation methods to create Winter's relation instances.

Winter's relation name and defined constraints functionality will now be called in our own handler after the relation has been instantiated. This will allow us to review the relation directly before any Winter code runs.

bennothommo committed:
Add ability to define relations (through methods) as dependent.

This allows people using the relation method format to define the "delete" attribute available to relation definition arrays.

bennothommo committed:
Fix Stan issue

bennothommo committed:
Add array schema for relations, add custom return types for relation methods

bennothommo committed:
Fix relation name for attachments

bennothommo committed:
Allow dependent model to be marked as not dependent

bennothommo committed:
Shore up API, strengthen types, remove validation method

The "validateRelationArgs" method does a whole lot of nothing - the most it does is enforce required parameters. These can be picked up in Laravel anyway.

bennothommo committed:
Minor

bennothommo committed:
Add ability to define model as pushable, add all relation definitions

bennothommo committed:
Add ability to define detachable relations

bennothommo committed:
Add attachment tests

bennothommo committed:
Fix attachment relations

Added the ability to define the field name and use this as a constraint for attachment relations, separating it from the relation name.

bennothommo committed:
Copilot went a bit hog-wild on that one

bennothommo committed:
Fix Stan issue

bennothommo committed:
Rename MigratesForTest class

The "Test" suffix might inadvertently be picked up by PHPUnit as a test case.

bennothommo committed:
Move Sortable trait tests

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/laravel-style-relations

bennothommo committed:
Add extension support for relations, add support for SoftDelete trait

bennothommo committed:
Merge branch 'develop' into wip/laravel-style-relations

bennothommo committed:
Bring in all remaining tests from core

bennothommo committed:
Fix soft deletion

bennothommo committed:
Add soft delete tests specifically with Laravel relations

bennothommo commented:
This PR is now ready for review. Will submit a PR to the docs that will indicate the API changes.

mjauvin commented:
Wow, that is quite extensive... Nice work!

How confident are you in the unit tests? ;)

bennothommo commented:
@mjauvin I've basically added copies of nearly every test we had to directly test the Laravel-style relations, so pretty confident :)

mjauvin commented:
So you're confident about the winter style relations as well? There were a LOT of changes...

bennothommo commented:
@mjauvin the original tests for the array-style relations are still there :)

bennothommo commented:
After giving this some thought, this might need to be released on a new "major" version update since we're not using semver.

The ramification of using this functionality - especially inside of a plugin - would lock that to the version of Winter that introduced it. So while it doesn't break backwards compatibility of the public API of Winter as it is now (ie. it doesn't cause any current functionality to fail), as soon as someone uses it, it immediately makes their Winter CMS install or plugin incompatible with older versions of Winter.

LukeTowers commented:
@bennothommo I'm fine with that, it'll provide some motivation to make the jump to Laravel 10 or 11 😉

bennothommo committed:
Merge branch 'develop' into wip/laravel-style-relations

bennothommo committed:
Merge branch 'fix-through-relations' into wip/laravel-style-relations

Also add ability to define relations as "counted" relations only.

LukeTowers commented:
@bennothommo I think I don't care about plugins choosing to become incompatible with older versions of Winter. Plugins are free to use semver and can define their requirements of a specific winter version in their composer.json; so I don't think this needs to wait for a new major version.

What is needed to get this across the line?

bennothommo commented:
@LukeTowers we'd just need to fix the merge conflict and tests. Otherwise, it's functionally complete.

bennothommo committed:
Merge branch 'develop' into wip/laravel-style-relations

bennothommo committed:
Fix database test cases not always running migrations

bennothommo committed:
Fix count-only relations.

When using the Laravel style relation method and countOnly flag, it was not applying the constraint automatically.

bennothommo committed:
Fix definition

bennothommo committed:
Apply count constraint directly on relation

bennothommo committed:
Fix Stan issues

bennothommo committed:
Fix support for pivot models

LukeTowers commented on src/Database/Attributes/Relation.php:
```diff
@@ -0,0 +1,10 @@
+#[Attribute(Attribute::TARGET_METHOD)]
+class Relation
+{
+}
```
What is this used for?

bennothommo replied:
This attribute can be used instead of a Relation class return type to indicate a method is a relation method. It's to aid discovery of relations in a model, since some functionality of Winter requires knowledge of the relationship configuration without loading the relation directly (such as the soft delete trait needing to know relations of the related models in order to cascade soft deletions).

It could be used in cases where a relation is dynamic or if, for some reason, the dev doesn't want to use return types.

LukeTowers commented on src/Database/Relations/Concerns/DefinedConstraints.php:
```diff
@@ -130,4 +117,16 @@ public function addDefinedConstraintsToQuery($query, ?array $args = null)
+    /**
+     * Get the relation definition for the related model.
+     *
+     * @return array
```
```suggestion
```

LukeTowers commented on src/Database/Relations/Concerns/DefinedConstraints.php:
```diff
@@ -11,20 +14,18 @@ trait DefinedConstraints
      * Set the defined constraints on the relation query.
      *
+     * This method is kept for backwards compatibility, but is no longer being called directly by Storm when
+     * initializing relations. Constraints are now applied on an as-needed basis.
```
Where are they being applied on the as-needed basis? Also should we tag this as deprecated if we're not going to use it anymore?

bennothommo replied:
This method used to pull double functionality. It set constraints on the relation object, and set constraints on the query, but this method called both on the relation object itself when retrieving the relation.

Setting constraints on the relation now is only needed when using property style relations, because in method style relations, you would add the constraints directly in the method, such as conditions, ordering, default models, timestamps, etc.

The query constraints are also only automatically applied on property style relations, and only when needed in the `Winter\Storm\Database\Concerns\HasRelationships` concern. I also saw it being used in the Relation Controller, so I kept that method available.

Thus, we no longer need this method - however, I couldn't discount that it might be used in a plugin somewhere.

bennothommo replied:
Oh, and yes, we should mark it deprecated

LukeTowers commented on src/Database/Relations/Concerns/DefinedConstraints.php:
```diff
@@ -60,26 +64,6 @@ public function addDefinedConstraintsToRelation($relation, ?array $args = null)
-                ->groupBy($foreignKey)
-                ->orderBy($foreignKey)
-            ;
-        }
```
Where is this handled now?

bennothommo replied:
In the `Winter\Storm\Database\Relations\Concerns\CanBeCounted` concern. :)

LukeTowers commented on src/Database/Relations/Concerns/CanBePushed.php:
````diff
@@ -0,0 +1,69 @@
+ *
+ * ```php
+ * public $hasMany = [
+ *     'messages' => [Message::class, 'push' => true]
````
```suggestion
 *     'messages' => [Message::class, 'push' => false]
```

bennothommo replied:
This example is to show how to enable push when using the property style, so it would need to be true.

LukeTowers commented on src/Database/Relations/Concerns/CanBePushed.php:
```diff
@@ -0,0 +1,69 @@
+namespace Winter\Storm\Database\Relations\Concerns;
+
+/**
+ * This trait is used to mark certain relationships as pushable.
```
Is this something new from Laravel?

bennothommo replied:
I believe it's been in Laravel for quite a while, and it's something that we support in Winter through the `push` option (see [our docs](https://wintercms.com/docs/v1.2/docs/database/relations#detailed-definitions) for more detail).

LukeTowers commented on src/Database/Relations/Concerns/CanBeExtended.php:
```diff
@@ -0,0 +1,126 @@
+ * Adds extension support for relations.
+ *
+ * Since Laravel allows macros for relations, we should also provide Winter's extension framework for relations, to
+ * allow scenarios like model traits being able to define custom relation methods for configuring the relation.
```
Could you add some examples here for what this change enables in Winter now, perhaps alongside the Laravel macros support? Trying to wrap my head around it.

bennothommo replied:
An example is for the `SoftDelete` trait - https://github.com/wintercms/storm/blob/c4abef88cb1c9909395e3d8e4cc176775e08d35b/src/Database/Traits/SoftDelete.php#L79-L112

I needed to allow for relations to be able to add chainable methods and additional properties for relation method definitions when circumstances require it, and Laravel macros don't allow you to define properties, only methods. In the case here with the `SoftDelete` trait, it is dynamically creating the "flag" methods to add a chained call to a relation method to define it as soft-deleteable.

LukeTowers replied:
Does this get run only once globally or is it called every time the SoftDelete trait is booted? @bennothommo

bennothommo replied:
Good question - I'll need to check this. It might be every time it's booted.

bennothommo replied:
Added a check to see if the dynamic methods and properties are defined already, and skip their redefinition if so.

LukeTowers commented on src/Database/Relations/Concerns/CanBeDetachable.php:
```diff
@@ -0,0 +1,71 @@
+namespace Winter\Storm\Database\Relations\Concerns;
+
+/**
+ * This trait is used to mark certain relationships as detachable.
```
Is this something new from Laravel?

bennothommo replied:
Nope, it's been available to Laravel for a while. We also support it through the `detach` property in the current property-style definition.

LukeTowers commented on src/Database/Relations/Concerns/CanBeDependent.php:
```diff
@@ -0,0 +1,73 @@
+namespace Winter\Storm\Database\Relations\Concerns;
+
+/**
+ * This trait is used to mark certain relationships as dependent.
```
This one already existed in Winter under the `delete` option right?

bennothommo replied:
That's right, but I didn't like the term deleteable or canDelete. It didn't quite convey what the functionality does.

LukeTowers commented on src/Database/Relations/Concerns/CanBeCounted.php:
````diff
@@ -0,0 +1,90 @@
+ * the relationship definition method. For example:
+ *
+ * ```php
+ * public function messages()
````
```suggestion
 * public function totalMessages()
```

LukeTowers commented on src/Database/Relations/Concerns/CanBeCounted.php:
````diff
@@ -0,0 +1,90 @@
+ *
+ * ```php
+ * public $hasMany = [
+ *     'messages' => [Message::class, 'count' => true]
````
```suggestion
 *     'total_messages' => [Message::class, 'count' => true]
```

LukeTowers commented on src/Database/Relations/Concerns/AttachOneOrMany.php:
```diff
@@ -12,14 +12,14 @@ trait AttachOneOrMany
+     * The field name (relation) to associate this attachment with.
      */
-    protected $public;
+    protected string $fieldName;
```
Why was the order of these properties switched? Messier diff this way. Also what exactly is the difference between relationName and fieldName?

bennothommo replied:
The position wasn't switched - it was simply that the `$relationName` property was removed, and the `$fieldName` property as added afterwards. `$relationName` is now specified in the `Winter\Storm\Database\Relations\Concerns\HasRelationName` concern.

`$fieldName` is required for parity with Laravel's base morph relations, although I don't believe we use it ourselves.

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+    use Concerns\CanBeCounted;
+    use Concerns\CanBeDetachable;
+    use Concerns\CanBeExtended;
+    use Concerns\CanBePushed;
```
Thoughts on the `CanBeX` vs `Xable` naming scheme?

bennothommo replied:
Subjective, but I feel this style reads better. ie. relation type can be counted, can be detached, etc.

LukeTowers replied:
Do we have other examples in our code (or Laravel's) using the CanBe? Just wondering if it's worth breaking convention.

bennothommo replied:
Laravel uses the prefix for its concerns sometimes, such as [`CanBeOneOfMany`](https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Eloquent/Relations/Concerns/CanBeOneOfMany.php) or [`CanBePrecognitive`](https://github.com/laravel/framework/blob/11.x/src/Illuminate/Http/Concerns/CanBePrecognitive.php). I think it's a good way to describe a concern that provides functionality that may be optional.

LukeTowers replied:
Good enough for me!

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+         */
+        if ($value instanceof Model) {
+            $value = $value->getKey();
+        }
+        elseif (is_array($value)) {
```
```suggestion
        } elseif (is_array($value)) {
```
We should probably make some adjustments to our published coding style and apply it in bulk across our codebases soon.

bennothommo replied:
We do cover it under **Curly braces for condition blocks** here: https://wintercms.com/docs/v1.2/docs/architecture/developer-guide#php-coding-standards

It's probable that I missed this or copied it from some old code.

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+            }
+        }
+
+        /*
+         * Convert scalar to array
+         */
```
```suggestion
        // Convert scalar to array
```

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+            $value = [$value];
+        }
+
+        /*
+         * Setting the relationship
+         */
```
```suggestion
        // Setting the relationship
```

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+            return;
+        }
+
+        /*
+         * Convert models to keys
+         */
```
```suggestion
        // Convert models to keys
```

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+    {
+        $relationModel = $this->getRelated();
+
+        /*
+         * Nulling the relationship
+         */
```
```suggestion
        // Nulling the relationship
```

LukeTowers commented on src/Database/Concerns/HasRelationships.php:
```diff
@@ -279,520 +338,573 @@ protected function getRelationDefaults($type)
+                \Winter\Storm\Database\Relations\Concerns\CanBeDependent::class,
+                class_uses_recursive($relation)
+            )
+            && (($definition['delete'] ?? false) === true)
```
Should we check the definition configs before making calls to reflection?

bennothommo replied:
Fixed.

LukeTowers commented on src/Database/Concerns/HasRelationships.php:
```diff
@@ -279,520 +338,573 @@ protected function getRelationDefaults($type)
+     * Finds the calling function name from the stack trace.
      */
-    public function hasOne($related, $primaryKey = null, $localKey = null, $relationName = null)
+    protected function getRelationCaller(): ?string
```
What is this being used for?

bennothommo replied:
It's used to determine the relation name for relations created by calling a method when you've only defined property-style relations. It was in the original version of this file.

Basically, it's so, for example:

```
$comments = $post->comments()->get();
```

When calling `comments()`, if that was defined as a property relation, it uses `comments` as the relation name when creating the relation object, and determining which config to retrieve.

LukeTowers commented on src/Database/Concerns/HasRelationships.php:
```diff
@@ -279,520 +338,573 @@ protected function getRelationDefaults($type)
+     * {@inheritDoc}
      */
-    public function morphedByMany($related, $name, $table = null, $primaryKey = null, $foreignKey = null, $parentKey = null, $relatedKey = null, $relationName = null)
+    protected function newMorphMany(Builder $query, Model $parent, $type, $id, $localKey)
```
Are these `new$relationType` methods in the same order as the old ones? Just wondering in case we could get a cleaner diff for them. Also is there any concerns with the public methods being removed?

bennothommo replied:
There's two changes happening here:

1. The original public methods are no longer in our `HasRelationships` definition. These are now provided by Laravel core.
2. The protected `new$relationType` methods are what we extend from the original Laravel definitions, mainly to provide a relation name and some small adjustments for our own needs. I believe these were *not* in the original file, so that's the reason for the messy diff.

LukeTowers commented on src/Database/Model.php:
```diff
@@ -849,12 +849,10 @@ public function newPivot(EloquentModel $parent, array $attributes, $table, $exis
-            $pivotModel = $definition['pivotModel'];
-            return $pivotModel::fromAttributes($parent, $attributes, $table, $exists);
-        }
+        return $pivotModel::fromAttributes($parent, $attributes, $table, $exists);
```
```suggestion
        return $pivotModel::fromRawAttributes($parent, $attributes, $table, $exists);
```

See https://github.com/wintercms/storm/commit/339af87267d355a8acef6dafbed9d2f2523099d2

LukeTowers commented on src/Database/Relations/BelongsToMany.php:
```diff
@@ -1,10 +1,147 @@
+            $related = $this->getRelated();
+
+            $value = $this->parent->getRelation($this->relationName)->pluck($related->getKeyName())->all();
+        }
+        else {
```
```suggestion
        } else {
```

bennothommo committed:
Update src/Database/Relations/Concerns/DefinedConstraints.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/Concerns/CanBeCounted.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/Concerns/CanBeCounted.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/BelongsToMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/BelongsToMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/BelongsToMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/BelongsToMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/BelongsToMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Model.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/BelongsToMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Merge branch 'develop' into wip/laravel-style-relations

LukeTowers commented on src/Database/Relations/Concerns/DefinedConstraints.php:
```diff
@@ -11,20 +14,18 @@ trait DefinedConstraints
     /**
      * Set the defined constraints on the relation query.
      *
+     * This method is kept for backwards compatibility, but is no longer being called directly by Storm when
+     * initializing relations. Constraints are now applied on an as-needed basis.
```
```suggestion
     * @deprecated 1.3.0 Constraints are now applied on an as-needed basis.
```

bennothommo committed:
Update src/Database/Relations/Concerns/DefinedConstraints.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Merge branch 'develop' into wip/laravel-style-relations

bennothommo committed:
Optimise conditions for applying traits and constraints to relation

bennothommo committed:
Fix Stan issues

bennothommo committed:
Stop double-defining of soft delete trait's dynamic methods and properties

LukeTowers commented on src/Database/Traits/SoftDelete.php:
```diff
@@ -62,6 +75,46 @@ public static function bootSoftDelete()
+                // Prevent double-defining the dynamically added properties and methods below
+                if ($this->methodExists('softDeletable')) {
+                    return;
+                }
```
This still concerns me as I think it becomes an n+1 issue where if a model uses SoftDelete then every instance of that model will cause an additional function to be added to these relation classes, even if that function exits early. Thoughts @bennothommo @jaxwilko?

bennothommo replied:
It would be minimised because it would be only on each "class" that uses it, not each model instance. But otherwise, I'm running out of ideas. I tried doing a "local" extension but I think that's too late in the boot process for it to be available in relation methods.

LukeTowers reviewed (changes requested):
Still have performance issues with the softdelete trait

bennothommo committed:
Use alternative method for defining soft deletable relations

Instead of extending the Relation class, we will make the soft delete methods available immediately to the relations.

However, you can only *enable* soft delete if the related model uses the soft delete trait, thereby ensuring the correct soft delete functionality is available to the model when required.

bennothommo committed:
Stan fixes

bennothommo commented:
@LukeTowers I've changed the soft delete methods on the relation to be available at all times, but the actual soft delete flag will only be enabled when `softDeletable()` is called on the relation AND the related model has the `SoftDelete` trait.

LukeTowers commented on src/Database/Dongle.php:
```diff
@@ -65,10 +65,6 @@ public function parse($sql)
         $result = preg_replace_callback('/group_concat\((.+)\)/i', function ($matches) {
-            if (!isset($matches[1])) {
-                return $matches[0];
-            }
```
Why was this required?

bennothommo replied:
It's pointless. The first match will always be present if the callback is run, because the callback won't be run if there's no expression match.

LukeTowers commented on src/Html/Helper.php:
```diff
@@ -35,10 +35,6 @@ public static function nameToArray($string)
         if (preg_match('/^([^\]]+)(?:\[(.+)\])+$/', $string, $matches)) {
-            if (count($matches) < 2) {
-                return $result;
-            }
```
Why was this changed?

bennothommo replied:
It'll never be fired. There will always be more than 2 matches.

LukeTowers commented:
@bennothommo is there a related docs PR?

bennothommo commented:
@LukeTowers yep, here: https://github.com/wintercms/docs/pull/176

bennothommo committed:
Merge branch 'develop' into wip/laravel-style-relations

bennothommo committed:
Add support for dynamic Laravel-style relations through Extension

bennothommo committed:
Stan fix

bennothommo committed:
Fix infinite loop, improve support for dynamic Laravel relations, improve performance

To prevent a particular infinite loop, we need to define the Relation type when using the Relation attribute for a method so that we don't have to instantiate the relation in order to find out what type it is.

bennothommo committed:
Stan fix

bennothommo commented:
Late change: Using the `Relation` attribute to define a relation method will now require the relation type to be defined as the attribute's first parameter. Only a minor API change, but critical for behind-the-scenes to prevent infinite loops.

bennothommo committed:
Allow relation flags to be toggled

bennothommo committed:
Minor

bennothommo committed:
Detect relation conflicts

When a relation is defined in both the relation properties and as a method with the same name, an exception will be thrown.

bennothommo committed:
Fix code smell

LukeTowers commented on src/Database/Concerns/HasRelationships.php:
```diff
@@ -14,8 +20,27 @@
 use Winter\Storm\Database\Relations\MorphToMany;
-use Winter\Storm\Support\Str;
-
+use Winter\Storm\Exception\ApplicationException;
```
SystemException instead please

bennothommo committed:
Use SystemException rather than ApplicationException

bennothommo committed:
Fix tests
